coderedcms/models/page_models.py

```diff
@@ -1053,3 +1067,268 @@ def serve_preview(self, request, mode):
+        CoderedWebPage.settings_panels + 
+        [
+            MultiFieldPanel(
+                [
+                    FieldPanel('auto_update_latlng'),
+                    FieldPanel('latitude'),
+                    FieldPanel('longitude'),
+                ],
+                heading=_("Location Settings")
+            ),
+        ]
+    )
+
+    @property
+    def get_geojson_name(self):
+        return self.map_title or self.title
+
+    @property
+    def get_geojson_description(self):
+        return self.map_description
+
+    @property
+    def render_pin_description(self):
+        return mark_safe(
+            """
+                <b>{0}</b>
+                <p>{1}</p>
+                <p><a href='{2}''>View Location</a></p>
+            """
+            .format(
+                self.title,
+                self.address,
+                self.url
+            )
+        )
+
+    @property
+    def render_list_description(self):
+        return mark_safe(
+            """
+            <div class='list-group-item flex-column align-items'>
+                <div class='d-flex w-100 justify-content-between'>
+                    <a href='{0}'><b class='mb-1'>{1}</b></a>
+                </div>
+                <small>{2}</small>
+                <br />
+                <small>{3}</small>
+            </div>
+            """
+            .format(
+                self.url,
+                self.get_geojson_name,
+                self.address,
+                self.get_geojson_description
+            )
+
+        )
+    
+    def to_geojson(self):
+        return {
+            "type": "Feature",
+            "geometry":{
+                "type": "Point",
+                "coordinates": [self.longitude, self.latitude]
+            },
+            "properties":{
+                "list_description": self.render_list_description,
+                "pin_description": self.render_pin_description
+            }
+        }
+
+
+    def serve(self, request, *args, **kwargs):
+        data_format = request.GET.get('data-format', None)
+        if data_format == 'geojson':
+            return self.serve_geojson(request, *args, **kwargs)
+        return super().serve(request, *args, **kwargs)
+
+    def serve_geojson(self, request, *args, **kwargs):
+        return JsonResponse({
+            "type": "FeatureCollection",
+            "features": [
+                self.to_geojson()
+            ]
+        })
+
+    def save(self, *args, **kwargs):
+        if self.auto_update_latlng and GoogleApiSettings.for_site(Site.objects.get(is_default_site=True)).google_maps_api_key:
+            g = geocoder.google(self.address, key=GoogleApiSettings.for_site(Site.objects.get(is_default_site=True)).google_maps_api_key)
+            try:
+                self.latitude = g.latlng[0]
+                self.longitude = g.latlng[1]
+            except TypeError:
+                """Raised if google denied the request"""
+                pass
```

Comment:
Shouldn't `geocoder.google()` go inside the try/except? i.e. in event of api timeout or someting. Also, I would think a message should be shown. Similar to wagtail's error/success messages on page save. Probably a "warning" level would do. E.g. "The address could not be geocoded".

Reply:
An error won't be raised if it times out.  The geocoder would just return an object with timed-out request.  I'm not entirely sure if we are able to add a message in this context, but I will try.

Reply:
Ok. I would still feel better putting in in the try block, for any future unforeseen breakage, especially since it is a 3rd party package. To prevent people from not being able to save their pages if it happens to break.